### ISSUE TITLE
Residue template issues

### ISSUE THREAD

**Issue by andrewsb8** [openmm/openmm#3413]
Hi,

I'm new to OpenMM but use Gromacs and python frequently so I'm not completely new to all of this. I am learning OpenMM primarily to use the CHARMM Drude 2019 polarizable force field. To start testing, I am taking pdb's generated from a simulation in CHARMM36m by Gromacs' trjconv tool and just trying to initialize a system to prepare for equilibration in Charmm36 (is the forcefield here updated to be the C36m parameters?) with OpenMM. I am only running the following commands:

```from openmm import *
from openmm.app import *
from openmm.unit import *
from sys import stdout
import sys

#inputs
protein = sys.argv[1]
ff = sys.argv[2]
water = sys.argv[3]

#import structure and create system
try:
    pdb = PDBFile(protein)
except ZeroDivisionError:
    print("If your pdb has a line like:\nCRYST1 0 0 0...\nat the top, delete it and run this script again.")

forcefield = ForceField(ff, water)

system = forcefield.createSystem(pdb.topology, nonbondedMethod=PME,
        nonbondedCutoff=1*nanometer, constraints=HBonds)

```
For the ff and water command line inputs, I am using charmm36.xml and charmm36/water.xml. This works (or at least doesn't throw an error) for the Villin structure input.pdb in the examples directory. However, if I use a pdb generated from my simulations of glycine-isoleucine-glycine ([GIG_ILE2_tempErr.txt](https://github.com/openmm/openmm/files/7871213/GIG_ILE2_tempErr.txt), extension changed to txt so I could upload), then I get an error:

 ```ValueError: No template found for residue 2 (ILE).  This might mean your input topology is missing some atoms or bonds, or possibly that you are using the wrong force field. ```

This is a standard residue and you can see that all of the hydrogens are included in the pdb, generated automatically by gromacs, so I shouldn't need to initialize a Modeller to add the hydrogens and I'm using the same forcefield, eliminating the possibility of naming conventions being weird. This issue gets worse, though. I thought maybe I would import a solvated, neutralized glycine-isoleucine-glycine system to better mimic the Villin example structure. This pdb ([GIG_GLY3_tempErr.txt](https://github.com/openmm/openmm/files/7871241/GIG_GLY3_tempErr.txt)), of the same structure plus water/ion, throws a different error:

 ```ValueError: No template found for residue 3 (GLY).  The set of atoms is similar to GLYN, but it is missing 1 hydrogen atoms. ```

Both pdbs are generated from the same trajectory in Gromacs with all hydrogens, so I'm unsure why one file is getting an ILE and one is getting a GLY template error. But, I am using NH3+ and COOH terminal cappings and I saw the Villin example using NH3+ and COO- cappings. So I deleted the last hydrogen of GLY3, resulting in COO-, in this pdb and I **don't** get an error! The problem is then:

- I need the neutral C terminal capping, which I know is compatible with CHARMM36 and CHARMM36m. But that doesn't seem to be working?
- Deleting the C terminal neutralizing hydrogen in the first of the two pdbs I have attached does not eliminate the ILE template error. So this is inconsistent behavior for systems where the only difference is the presence of water and a box definition.
- I am getting varying error messages for the same system and I genuinely cannot figure out why when I am using standard pdb files automatically generated.
- As a sanity check, I also tried just generating a structure for GIG in VMD and just loading it. Still getting an GLY3 template error, so I guess it's not an issue with Gromacs' trjconv?

I do apologize for the novel and greatly appreciate any help. I hope that I'm not just doing something terribly wrong.

Thanks,

Brian

**Comment by peastman** [comment]
The first problem is in your element specifications:

```
ATOM     17 1HG2 ILE X   2      22.162  19.247  21.479  1.00  0.00          HG
ATOM     18 2HG2 ILE X   2      22.582  20.167  19.965  1.00  0.00          HG
ATOM     19 3HG2 ILE X   2      23.322  20.467  21.579  1.00  0.00          HG
```

The last column is supposed to contain the atomic symbol for the element.  You've told it those atoms are mercury, not hydrogen!

The second problem is in the name of the terminal hydrogen:

```
ATOM     37  HT2 GLY X   3      18.922  14.707  22.019  1.00  0.00           H
```

The standard name for that atom is HXT.  With a nonstandard name, the PDB loader doesn't know what it should be bonded to.  See [the FAQ](https://github.com/openmm/openmm/wiki/Frequently-Asked-Questions#template) for more information about this.

**Comment by andrewsb8** [comment]
Hi @peastman 

Thank you for your help. Those modifications did eliminate the template errors I was receiving. The explanation in the FAQ makes sense, given that is also the case in other MD software. I definitely did assume that the output atom types using atom names from the CHARMM36 forcefield with Gromacs would automatically match between software packages. I will definitely take more care as I am preparing more systems in the future.

Thanks again,

Brian

**Comment by andrewsb8** [comment]
May have spoken too soon. Regarding the C terminal glycine. I used the following line

`ATOM     37  HXT2GLY X   3      18.922  14.707  22.019  1.00  0.00         HXT`

which is the only thing that removes the template error. But I print a pdb after solvation and this atom has been removed from the system, resulting in a COO- C terminus. I thought maybe I messed up the format of that line in the pdb, but everything else I've tried throws a template error again for that line.

Ctrl+F for "HXT" shows no results in the charmm36.xml file. Similarly, the only GLY records (that are glycine and not some other molecule like GLYC) are GLY/GLYM which show a non-terminal glycine residue (NH n terminus and CO C terminus, and GLYN which seems to indicate a COOH C terminus but requires two hydrogens bonded to the Nitrogen. Not sure if I am continuing to miss something. Thanks again for your help!

Brian

**Comment by peastman** [comment]
> `ATOM 37 HXT2GLY X 3 18.922 14.707 22.019 1.00 0.00 HXT`

Two issues with that line.  First, it's missing a lot of spaces.  Make sure they're present in the real file.  PDB files are based around fixed width columns.  Second, you've told it the element is "HXT", which of course is not the name of any element.  The element should be just "H".

> But I print a pdb after solvation and this atom has been removed from the system, resulting in a COO- C terminus.

It won't get removed unless you do something to remove it.  At what point does it disappear?  What function do you call that causes it to get removed?

> Ctrl+F for "HXT" shows no results in the charmm36.xml file.

The names that appear in force field definitions are just arbitrary labels.  They don't need to follow any particular naming convention.  The important point is the PDB reader.  It expects the file to follow standard PDB naming, so it can figure out where to add bonds to the Topology.

> Similarly, the only GLY records (that are glycine and not some other molecule like GLYC) are GLY/GLYM which show a non-terminal glycine residue

CHARMM uses patches to define terminal residues.  Inside the tag for GLY, you'll find a bunch of `<AllowPatch>` tags.  Those tell it that various patches can be applied to modify the residue.  The relevant ones here are CTER and CNEU, which respectively correspond to charged and neutral terminal residues.

**Comment by andrewsb8** [comment]
Yeah apparently the formatting didn't follow when I copied from the pdb file I am using. The correct number of spaces do exist in the pdb file.

What I needed to do was remove the `2` so that the third entry of the pdb read `HXT` instead of `HXT2`. Then the system conserved the neutral C terminus and added the appropriate ions to neutralize. Thank you for explaining things in such detail.
Very much appreciate all of your help!
